File: problems/leetcode/202-happy-number/solution.js
const seen = {};

function isHappy(n) {
    if (n === 1) return true;
    if (seen[n]) return false;
    seen[n] = true;
    n = +n
        .toString()
        .split("")
        .map((v) => v ** 2)
        .reduce((acc, v) => acc + v, 0);
    return isHappy(n, seen);
}

function isHappyWhile(n) {
    const seen = {};
    while (n !== 1) {
        if (seen[n]) return false;
        seen[n] = true;
        n = +n
            .toString()
            .split("")
            .map((v) => v ** 2)
            .reduce((acc, v) => acc + v, 0);
    }
    return true;
}

console.log(isHappy(19));
console.log(isHappy(10));

File: problems/leetcode/202-happy-number/solution.rs
struct Solution;

use std::collections::HashMap;
impl Solution {
    pub fn is_happy(mut n: i32) -> bool {
        let mut seen: HashMap<i32, bool> = HashMap::new();

        while n != 1 {
            if seen.contains_key(&n) {
                return false;
            }
            seen.insert(n, true);
            n = n.to_string().chars().fold(0, |mut acc, v| {
                let digit = v.to_digit(10).unwrap() as i32;
                acc += digit.pow(2);
                acc
            });
        }
        true
    }
}

fn main() {
    assert_eq!(Solution::is_happy(19), true);
    assert_eq!(Solution::is_happy(2), false);
}
